dap/adapter.go

```diff
@@ -0,0 +1,395 @@
+package dap
+
+import (
+	"context"
+	"encoding/json"
+	"sync"
+
+	"github.com/docker/buildx/build"
+	"github.com/google/go-dap"
+	gateway "github.com/moby/buildkit/frontend/gateway/client"
+	"github.com/pkg/errors"
+	"golang.org/x/sync/errgroup"
+)
+
+type Adapter struct {
+	srv *Server
+	eg  *errgroup.Group
+	cfg build.InvokeConfig
+
+	initialized   chan struct{}
+	started       chan error
+	configuration chan struct{}
+
+	evaluateReqCh chan *evaluateRequest
+
+	threads      map[int]*thread
+	threadsMu    sync.RWMutex
+	nextThreadID int
+}
+
+func New() *Adapter {
+	d := &Adapter{
+		initialized:   make(chan struct{}),
+		started:       make(chan error, 1),
+		configuration: make(chan struct{}),
+		evaluateReqCh: make(chan *evaluateRequest),
+		threads:       make(map[int]*thread),
+		nextThreadID:  1,
+	}
+	d.srv = NewServer(d.dapHandler())
+	return d
+}
+
+func (d *Adapter) Start(ctx context.Context, conn Conn) error {
+	d.eg, _ = errgroup.WithContext(ctx)
+	d.eg.Go(func() error {
+		return d.srv.Serve(ctx, conn)
+	})
+
+	<-d.initialized
+	return <-d.started
+}
+
+func (d *Adapter) Stop() error {
+	if d.eg == nil {
+		return nil
+	}
+
+	d.srv.Go(func(c Context) {
+		c.C() <- &dap.TerminatedEvent{
+			Event: dap.Event{
+				Event: "terminated",
+			},
+		}
+		// TODO: detect exit code from threads
+		// c.C() <- &dap.ExitedEvent{
+		// 	Event: dap.Event{
+		// 		Event: "exited",
+		// 	},
+		// 	Body: dap.ExitedEventBody{
+		// 		ExitCode: exitCode,
+		// 	},
+		// }
+	})
+	d.srv.Stop()
+
+	err := d.eg.Wait()
+	d.eg = nil
+	return err
+}
+
+func (d *Adapter) Initialize(c Context, req *dap.InitializeRequest, resp *dap.InitializeResponse) error {
+	close(d.initialized)
+
+	// Set capabilities.
+	resp.Body.SupportsConfigurationDoneRequest = true
```

Comment:
`SupportsTerminateRequest` and `SupportsRestartRequest` seem to be needed to allow the terminate request and the restart request (c.f. https://microsoft.github.io/debug-adapter-protocol//specification.html )